src/transfer-manager.ts

```diff
@@ -0,0 +1,287 @@
+/*!
+ * Copyright 2022 Google LLC. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Bucket, UploadOptions, UploadResponse} from './bucket';
+import {
+  DownloadCallback,
+  DownloadOptions,
+  DownloadResponse,
+  File,
+} from './file';
+import * as pLimit from 'p-limit';
```

Comment:
`p-limit` is the library that will aid in limiting the number of executing promises and was already a dependency. If we really feel strongly against it, it is probably possible to roll our own. However, I'm not sure it would be worth the investment over a widely used / popular library.